pkg/ruler/ruler.go

```diff
@@ -544,68 +582,86 @@ func (r *Ruler) syncRules(ctx context.Context, reason string) {
 		r.ruleGroupSyncDuration.Set(ruleGroupSyncDuration)
 	}()
 
-	loadedConfigs, err := r.loadRuleGroups(ctx)
+	loadedConfigs, backupConfigs, err := r.loadRuleGroups(ctx)
 	if err != nil {
 		return
 	}
 
 	// This will also delete local group files for users that are no longer in 'configs' map.
 	r.manager.SyncRuleGroups(ctx, loadedConfigs)
+
+	if r.cfg.APIEnableRulesBackup {
+		r.manager.BackUpRuleGroups(ctx, backupConfigs)
+	}
 }
 
-func (r *Ruler) loadRuleGroups(ctx context.Context) (map[string]rulespb.RuleGroupList, error) {
+func (r *Ruler) loadRuleGroups(ctx context.Context) (map[string]rulespb.RuleGroupList, map[string]rulespb.RuleGroupList, error) {
 	timer := prometheus.NewTimer(nil)
 
 	defer func() {
 		storeLoadSeconds := timer.ObserveDuration().Seconds()
 		r.ruleGroupStoreLoadDuration.Set(storeLoadSeconds)
 	}()
 
-	configs, err := r.listRules(ctx)
+	ownedConfigs, backupConfigs, err := r.listRules(ctx)
 	if err != nil {
 		level.Error(r.logger).Log("msg", "unable to list rules", "err", err)
-		return nil, err
+		return nil, nil, err
 	}
 
-	loadedConfigs, err := r.store.LoadRuleGroups(ctx, configs)
+	loadedOwnedConfigs, err := r.store.LoadRuleGroups(ctx, ownedConfigs)
 	if err != nil {
-		level.Warn(r.logger).Log("msg", "failed to load some rules owned by this ruler", "count", len(configs)-len(loadedConfigs), "err", err)
+		level.Warn(r.logger).Log("msg", "failed to load some rules owned by this ruler", "count", len(ownedConfigs)-len(loadedOwnedConfigs), "err", err)
 	}
-	return loadedConfigs, nil
+	if r.cfg.APIEnableRulesBackup {
+		loadedBackupConfigs, err := r.store.LoadRuleGroups(ctx, backupConfigs)
+		if err != nil {
+			level.Warn(r.logger).Log("msg", "failed to load some rules backed up by this ruler", "count", len(backupConfigs)-len(loadedBackupConfigs), "err", err)
+		}
+		return loadedOwnedConfigs, loadedBackupConfigs, nil
+	}
+	return loadedOwnedConfigs, nil, nil
 }
 
-func (r *Ruler) listRules(ctx context.Context) (result map[string]rulespb.RuleGroupList, err error) {
+func (r *Ruler) listRules(ctx context.Context) (owned map[string]rulespb.RuleGroupList, backedUp map[string]rulespb.RuleGroupList, err error) {
```

Comment:
Our paginated rules api changed these files as well. should we reviewed on top of pagination apis' changes?